Script to convert sic-4 tables into divisions

Juan-Mateos opened the pull request:
Closes #45 

includes changes in `sic.py` to read all sic codes as `str`

Commit:
Script to convert sic-4 tables into divisions

Commit:
Script to convert sic-4 tables into divisions

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -0,0 +1,128 @@
+        A dict mapping source levels into target levels
+    """
+    lookup_table = (  # Each row in this lookup has a source and a target
+        sic[[target_level, source_level]].fillna(method="ffill").dropna(axis=0)
```
There is a bug here in the forward filling.
`source_level` ("Class") gets forward filled which propagates `source_level` beyond the next `target_level` marker.

E.g.

```
>>  sic[[target_level, source_level]].iloc[1290:1310].fillna(method="ffill")
Division	Class
1290	NaN	96.03
1291	NaN	96.04
1292	NaN	96.09
1293	NaN	96.09
1294	NaN	96.09
1295	NaN	96.09
1296	NaN	96.09
1297	97	96.09 <-- 96.09 Class gets labelled as Division 97
1298	97	96.09
1299	97	97.00
1300	97	97.00
1301	98	97.00
1302	98	97.00
1303	98	98.10
1304	98	98.10
1305	98	98.20
1306	98	98.20
1307	98	98.20
1308	98	98.20
1309	98	98.20
```

bishax replied:
Rather than fixing the forward fill, we can just take the first 2 characters of Class and then we get the division. We don't need to do it for generic `target_level` for this script anyway.

bishax replied:
To check the fix I went ahead and committed changes.

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -0,0 +1,128 @@
+    return table_reg
+
+
+def get_transform_nomis_data(
```
This function doesn't help made the code any clearer (it mixes getting and transforming the data)

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -0,0 +1,128 @@
+
+    NOMIS_FILE_LOC = f"{project_dir}/data/processed/nomis_divisions.csv"
+
+    if not os.path.exists(NOMIS_FILE_LOC):
```
This might cause errors if we update the underlying data (our `NOMIS_FILE_LOC` wouldn't get updated).
The script only takes about a second to run so we don't really need to avoid doing the work.

bishax committed:
Reviewer Changes

- FIX: `make_sic_target_lookup` had a bug (see review), fixed it and simplified to
    only make division lookup.
- Refactored script in the process of working through fix

bishax reviewed (commented):
There was a bug in the lookup - I'm about to commit a fix (and associated refactorings) for you to check and review.

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -1,128 +1,82 @@
+# %%
+"""Script to aggregate NOMIS data over SIC divisions."""
```
Tells us what the purpose of the file is:
- It's a script
- What it does

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -1,128 +1,82 @@
-    """
-    lookup_table = (  # Each row in this lookup has a source and a target
-        sic[[target_level, source_level]].fillna(method="ffill").dropna(axis=0)
+def make_sic_div_lookup(sic):
```
Fixes the bug flagged in last review.

For simplicity it restricts itself to create a division lookup (If we wanted something generic, we should focus the logic in a function that fills in all gaps in the spreadsheet)

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -1,128 +1,82 @@
 
 
-def convert_sic_data(
+def aggregate_NOMIS_over_SIC(
```
Pull out the main data operation into a separate function.

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -1,128 +1,82 @@
+    """Aggregates NOMIS table from SIC4 (Class) to `target_level`.
 
-    # Extract division from SIC4s
-    table_[target_level] = table_[source_level].map(source_target_lookup)
```
This can happen at a higher-level in the code (then we don't nave to pass `source_target_lookup` down two levels of functions)

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -1,128 +1,82 @@
-    table_reg[f"{target_level}_description"] = table_reg[target_level].map(
-        code_descr_lookup
-    )
-    table_reg["source"] = name
```
Again, this can happen at a higher level.

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -1,128 +1,82 @@
     """
-    # Get and subset the data
-    d = getter()
-    d = d.loc[d["year"].isin(years)]
```
Can happen at a higher level, then we don't need this "intermediate" function.

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -1,128 +1,82 @@
 
 
 if __name__ == "__main__":
+    # Ensure SIC data is downloaded
+    if not os.path.exists(_SIC_OUTPUT_FILE):
+        save_sic_taxonomy()
```
Get the data we depend on

bishax commented on sg_covid_impact/make_sic_division.py:
```diff
@@ -1,128 +1,82 @@
-                t, years, source, div_code_description, sic_div_lookup
+    nomis_divs = []
+    for getter, source in zip([get_IDBR, get_BRES], ["idbr", "bres"]):
+        nomis_divs.append(
+            getter()
+            .loc[lambda x: x.year.isin(years)]
+            .assign(division=lambda x: x.SIC4.map(sic_div_lookup))
+            .pipe(aggregate_NOMIS_over_SIC, target_level="division")
+            .assign(
+                division_description=lambda x: x.division.map(div_code_description),
+                source=source,
```
Reads as one sequential piece of logic:
- get the data
- filter by year
- add division column
- aggregate over division
- add description and data source

We're not passing lots of arguments deep into functions so it's easier to see what things are doing.

bishax reviewed (commented):
Annotated reasons for changes

Juan-Mateos commented:
When running the new script I get:

```
Traceback (most recent call last):
  File "sg_covid_impact/make_sic_division.py", line 9, in <module>
    from sg_covid_impact.getters.nomis import get_BRES, get_IDBR
  File "/Users/jmateosgarcia/Desktop/projects/sg_covid_impact/sg_covid_impact/getters/nomis.py", line 2, in <module>
    from sg_covid_impact.utils.metaflow import flow_getter, cache_getter_fn
  File "/Users/jmateosgarcia/Desktop/projects/sg_covid_impact/sg_covid_impact/utils/metaflow.py", line 11, in <module>
    from filelock import FileLock
ModuleNotFoundError: No module named 'filelock'
```

Commit:
Added filelock to env yaml and changed nomis getter

Commit:
Added filelock to env yaml and changed nomis getter

Commit:
Merge branch 'dev' into 45_bres_idbr
